base/ryu/Ryu.jl

```diff
@@ -0,0 +1,95 @@
+    pos = writeshortest(buf, 1, x)
+    return unsafe_string(pointer(buf), pos-1)
+end
+
+"""
+    Ryu.writefixed(x, plus=false, space=false, hash=true, precision=-1, decchar=UInt8('.'))
+    Ryu.writefixed(buf::Vector{UInt8}, pos::Int, x, args...)
+
+Convert a float value `x` into a "fixed" size decimal string.
+This function allows achieving the `%f` printf format.
+Note the 2nd method allows passing in a byte buffer and position directly; callers must ensure the buffer has sufficient room to hold the entire decimal string.
+
+Various options for the output format include:
+  * `plus`: for positive `x`, prefix decimal string with a `'+'` character
+  * `space`: for positive `x`, prefix decimal string with a `' '` character; overridden if `plus=true`
+  * `hash`: whether the decimal point should be written, even if no additional digits are needed for precision
+  * `precision`: minimum number of significant digits to be included in the decimal string; extra `'0'` characters will be added for padding if necessary
+  * `decchar`: decimal point character to be used
+"""
+function writefixed(x::T, precision) where {T <: Base.IEEEFloat}
+    buf = Vector{UInt8}(undef, precision + neededdigits(T))
+    pos = writefixed(buf, 1, x, false, false, false, precision)
+    return unsafe_string(pointer(buf), pos-1)
+end
+
+"""
+    Ryu.writeexp(x, plus=false, space=false, hash=true, precision=-1, expchar=UInt8('e'), decchar=UInt8('.'))
+    Ryu.writeexp(buf::Vector{UInt8}, pos::Int, x, args...)
+
+Convert a float value `x` into a scientific notation decimal string.
+This function allows achieving the `%e` printf format.
+Note the 2nd method allows passing in a byte buffer and position directly; callers must ensure the buffer has sufficient room to hold the entire decimal string.
+
+Various options for the output format include:
+  * `plus`: for positive `x`, prefix decimal string with a `'+'` character
+  * `space`: for positive `x`, prefix decimal string with a `' '` character; overridden if `plus=true`
+  * `hash`: whether the decimal point should be written, even if no additional digits are needed for precision
+  * `precision`: minimum number of significant digits to be included in the decimal string; extra `'0'` characters will be added for padding if necessary
+  * `expchar`: character to use exponent component in scientific notation
+  * `decchar`: decimal point character to be used
+"""
+function writeexp(x::T, precision) where {T <: Base.IEEEFloat}
+    buf = Vector{UInt8}(undef, precision + neededdigits(T))
+    pos = writeexp(buf, 1, x, false, false, false, precision)
+    return unsafe_string(pointer(buf), pos-1)
+end
+
+function Base.show(io::IO, x::T) where {T <: Base.IEEEFloat}
+    if get(io, :compact, false)
+        x = round(x, sigdigits=6)
+    end
+    buf = Vector{UInt8}(undef, neededdigits(T))
+    pos = writeshortest(buf, 1, x)
+    GC.@preserve buf unsafe_write(io, pointer(buf), pos - 1)
```

Comment:
```suggestion
    write(io, resize!(bu, pos - 1))
```

although it also seems like `write(io, writeshortest(x))` would give the same result?

Reply:
`write(io, writeshortest(x))` would make an extra copy of the buffer though (when creating the string)